### PR TITLE
File Viewer Plugin

### PR DESCRIPTION
**Pull request by JokerJosh1234**
this plugin will add a button to supported files to view the contents (contents dont get loaded unless to user presses the button)

https://github.com/Vendicated/Vencord/assets/103841613/e4502fed-0316-4229-9c81-f475415282f5

in the video you can see how it works, if the file size is over 10MB the preview wont work therefore we tell the uesr that. the video only shows 3 of the 14 supported file types but they all work the same. there are no settings for this plugin.

### CODE REVIEW

**Review by MrDiamondDog — CHANGES_REQUESTED** [review]
Don't use DOM manipulation. Use patches and react components to add whatever you need

File: src/utils/constants.ts [hunk 1]
```diff
@@ -33,6 +33,10 @@ export interface Dev {
  * If you are fine with attribution but don't want the badge, add badge: false
  */
 export const Devs = /* #__PURE__*/ Object.freeze({
+    JokerJosh: {
+        name: "JokerJosh",
+        id: 791550177780563998n,
+    },
```
Comment (lewisakura) [inline src/utils/constants.ts:39]: please don't put yourself in the top of the list

File: src/plugins/fileViewer/index.tsx [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+        } else {
+            message.appendChild(createIframe(extensionTypeLinks[type](encodeURIComponent(url)), type));
+            element.classList.add("fv-expanded");
+        }
+
+        element.innerHTML = !element.classList.contains("fv-expanded") ? ShowIcon : HideIcon;
+    });
+
+    return div;
+}
```
Comment (lewisakura) [inline src/plugins/fileViewer/index.tsx:146]: use react

File: src/plugins/fileViewer/index.tsx [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+        minWidth: "317px",
+        maxHeight: type === "object" ? "370px" : "80vh",
+        minHeight: "50px",
+        borderRadius: "20px"
+    };
+
+    Object.assign(frame.style, styles);
+
+    return frame;
+}
```
Comment (lewisakura) [inline src/plugins/fileViewer/index.tsx:170]: use react

File: src/plugins/fileViewer/index.tsx [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+        Object.assign(Classes, ...findBulk.apply(null, [filters.byProps("attachment", "metadata"), filters.byProps("scrollerInner"), filters.byProps("fileNameLink"), filters.byProps("layerContainer"), filters.byProps("chat")]));
+
+        setInterval(id => {
+            if (!chatContainer) {
+                chatContainer = document.querySelector(`.${Classes.chat}`);
+            } else {
+                observer.observe(chatContainer, { childList: true, subtree: true });
+                clearInterval(id);
+            }
+        }, 200);
```
Comment (lewisakura) [inline src/plugins/fileViewer/index.tsx:194]: use a patch to get the chatbox

File: src/plugins/fileViewer/index.tsx [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+    name: "File Viewer",
+
+    description: "View files in discord, such as .pdf, .doc, .docx, .obj, etc",
+
+    authors: [Devs.JokerJosh],
+
+    patches: [],
+
+    start() {
+        Object.assign(Classes, ...findBulk.apply(null, [filters.byProps("attachment", "metadata"), filters.byProps("scrollerInner"), filters.byProps("fileNameLink"), filters.byProps("layerContainer"), filters.byProps("chat")]));
```
Comment (lewisakura) [inline src/plugins/fileViewer/index.tsx:185]: what?

File: src/plugins/fileViewer/index.tsx [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+    stop() {
+        observer.disconnect();
+    }
+});
+
+const buildIcon = (color, viewBox, path) => `
+    <svg class="${Classes.downloadButton}" aria-hidden="true" role="img" width="24" height="24" viewBox="${viewBox}">
+        <path fill="${color}" d="${path}"></path>
+    </svg>
+`;
```
Comment (lewisakura) [inline src/plugins/fileViewer/index.tsx:206]: use react

File: src/plugins/fileViewer/index.tsx [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+            body: (
+                <div>
+                    <p>File is too large.</p>
+                </div>
+            ),
+            confirmText: "Okay!",
+        });
+    });
+    return div;
+}
```
Comment (lewisakura) [inline src/plugins/fileViewer/index.tsx:119]: use react

File: src/plugins/fileViewer/index.tsx [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+                if (!chatContainer) {
+                    chatContainer = document.querySelector(`.${Classes.chat}`);
+                } else {
+                    observer.observe(chatContainer, { childList: true, subtree: true });
+                    clearInterval(id);
+                }
+            }, 200);
+        }
+    }
+};
```
Comment (lewisakura) [inline src/plugins/fileViewer/index.tsx:87]: my understanding of this is that it's meant to add a button to the message, so please use the chat buttons API to do this

**Review by lewisakura — CHANGES_REQUESTED** [review]

File: src/plugins/fileViewer/index.tsx (hunk 8 above)
Comment (JokerJosh1234) [reply src/plugins/fileViewer/index.tsx:87]: where can i find the chat buttons api? also i literally cannot figure out how to use patches, istg ive spent atleast 5 hours trying to figure it out and haven't got a single bit closer than when i started.

Comment (Scyye) [reply src/plugins/fileViewer/index.tsx:87]: look at another plugin that uses the chat buttons api, such as [ViewRaw](https://github.com/Vendicated/Vencord/blob/main/src/plugins/viewRaw/index.tsx)

Comment (Sqaaakoi) [reply src/plugins/fileViewer/index.tsx:87]: chat buttons api wouldn't even be suitable for this, this was for attachments

**Pull request closed without merging**